### ISSUE TITLE
Clarification about TorchVision Pretrained Model Encoders usage

### ISSUE THREAD

**Issue by tboo** [ludwig-ai/ludwig#3170]
Hi,

This is not really a bug but more like a question for clarification since I might not understand the documentation properly.

What I want to do is to test various encoder architectures for my image classification problem. So I was quite happy when I saw the TorchVision encoder zoo popping up in the newest release and I was hoping to have access to many more architectures compared to `v0.6` and older. I now wanted to try the TorchVision Pretrained Model Encoders (without loading the weights) and train from scratch. E.g. just utilizing the architecture since my data is not very close to ImageNet content at all. I thought I could do this with a yaml like that:

```
input_features:
- name: image_path
  type: image
  preprocessing:
      num_channels: 1
      height: 250
      width: 250
  encoder:
      type: resnet
      use_pretrained: false
      trainable: true
      model_variant: 34

output_features:
- name: class
  type: category

trainer:
  epochs: 50
  batch_size: auto
  early_stop: 10
  learning_rate: 0.004
  learning_rate_scheduler:
      reduce_on_plateau: 4
      reduce_on_plateau_patience: 4
      reduce_plateau_rate: 0.5
```

This should, so I thought, train a ResNet34 from scratch. However, the training does not start since there is a conflict between output and broadcast shapes. That's also the reason why I defined my image dimensions manually, which was never required before. I still get:

```
[1]<stderr>:  File "/home/ray/anaconda3/bin/ludwig", line 8, in <module>
[1]<stderr>:    sys.exit(main())
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/cli.py", line 172, in main
[1]<stderr>:    CLI()
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/cli.py", line 67, in __init__
[1]<stderr>:    getattr(self, args.command)()
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/cli.py", line 72, in train
[1]<stderr>:    train.cli(sys.argv[2:])
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/train.py", line 395, in cli
[1]<stderr>:    train_cli(**vars(args))
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/train.py", line 185, in train_cli
[1]<stderr>:    model.train(
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/api.py", line 535, in train
[1]<stderr>:    preprocessed_data = self.preprocess(
[1]<stderr>:  File "/home/ray/anaconda3/lib/python3.8/site-packages/ludwig/api.py", line 1428, in preprocess
[1]<stderr>:    raise RuntimeError(f"Caught exception during model preprocessing: {str(e)}") from e
[1]<stderr>:RuntimeError: Caught exception during model preprocessing: output with shape [1, 224, 224] doesn't match the broadcast shape [3, 224, 224]
```

What puzzles me is the shape conflict that has nothing to do with the actual dimensions of my data. I get the same error with the `resnext` encoder. Do I have to resize my data accordingly?

What works however (because it worked before) is using the old `_resnet_legacy` encoder with a yaml below. I assumed naively I can use the TorchVision encoders the same way. but I might be wrong there.

```
input_features:
- name: image_path
  type: image
  preprocessing:
      num_channels: 1
      height: 250
      width: 250
  encoder:
    type: _resnet_legacy
    dropout: 0.0
    output_size: 128
    activation: relu
    norm: null
    first_pool_kernel_size: null
    first_pool_stride: null
    use_bias: true
    bias_initializer: zeros
    weights_initializer: xavier_uniform
    norm_params: null
    num_fc_layers: 1
    fc_layers: null
    resnet_size: 50
    num_channels: null
    out_channels: 32
    kernel_size: 3
    conv_stride: 1
    batch_norm_momentum: 0.9
    batch_norm_epsilon: 0.001

output_features:
- name: class
  type: category

trainer:
  epochs: 50
  batch_size: auto
  early_stop: 10
  learning_rate: 0.004
  learning_rate_scheduler:
      reduce_on_plateau: 4
      reduce_on_plateau_patience: 4
      reduce_plateau_rate: 0.5
```

Any feedback for clarification is much appreciated. I am using the ludwig-ray-gpu:master docker container (Off topic: this container does not seem to contain `mlflow` and requires a manual installation to avoid an early training crash).

**Comment by tgaddair** [comment]
Hey @tboo, thanks for reporting this. The issue here is that the pretrained TorchVision models all assume 3 channel input images. I consider this a bug on our side, since it should work without you needing to do anything in particular to your data or adjust preprocessing.

At the moment I'm thinking the fix on our side should be to automatically replicate the single channel as 3 channels during preprocessing to feed into the TorchVision model (similar to how we crop to 224x224 during preprocessing to fit the input expectations). That way it would just work transparently without needing to modify data or preprocessing.

cc @jimthompson5802 

Regarding mlflow, you are correct that we don't include it my default, since it's considered a "contrib" module (so we don't list it in any of our requirements). However, this could be a reasonable thing to do since a lot of users are using mlflow. I can make a note to fix that.

**Comment by tgaddair** [comment]
Hey @tboo, we just release Ludwig v0.7.1 which contains the fix in #3173 for this issue. Going forward, all torchvision models in Ludwig should play nicely with 1 channel images, without needing to override any preprocessing parameters. @geoffreyangus can give you more details if you have questions.

Please try it out and let us know if you run into any issues!

**Comment by tboo** [comment]
Thanks a lot for the super fast fix. This works now as expected also with single channel images. Much appreciated.
